runtime/src/bank.rs

```diff
@@ -4185,7 +4186,11 @@ impl Bank {
     /// Register a new recent blockhash in the bank's recent blockhash queue. Called when a bank
     /// reaches its max tick height. Can be called by tests to get new blockhashes for transaction
     /// processing without advancing to a new bank slot.
-    pub fn register_recent_blockhash(&self, blockhash: &Hash) {
+    fn register_recent_blockhash(&self, blockhash: &Hash, scheduler: &InstalledSchedulerRwLock) {
+        // This is needed because recent_blockhash updates necessitate synchronizations for
+        // consistent tx check_age handling.
+        BankWithScheduler::wait_for_paused_scheduler(self, scheduler);
```

Comment:
this is indeed very subtle yet crucial. there's tests for this: `test_scheduler_schedule_execution_recent_blockhash_edge_case_with_race` in #33070

Reply:
So I'm trying to just summarize the reason this is necessary.

Essentially, in the loop in blcokstore_processor.rs, scheduling no longer blocks our transaction processing. This means we will end up scheduling until we hit the final tick, and which point we register the new bank hash.

We need to make sure the blockhash_queue of our bank is **not** updated until after all the scheduled transactions have completed, because if a tx w/ recent blockhash equal to the current blockhash was scheduled we'd get inconsistent results based on when it was **executed** not scheduled.

I think that is correct, but please let me know if it is not.
If that is the case, why did we choose this approach instead of checking blockhash at schedule time instead?
afaik blockhash queue is only updated at the end of a block, i.e. when we should not be scheduling any more txs for that slot.

Reply:
> Essentially, in the loop in blcokstore_processor.rs, scheduling no longer blocks our transaction processing. This means we will end up scheduling until we hit the final tick, and which point we register the new bank hash.
> 
> We need to make sure the blockhash_queue of our bank is **not** updated until after all the scheduled transactions have completed, because if a tx w/ recent blockhash equal to the current blockhash was scheduled we'd get inconsistent results based on when it was **executed** not scheduled.

that's correct understanding.

> If that is the case, why did we choose this approach instead of checking blockhash at schedule time instead?
> afaik blockhash queue is only updated at the end of a block, i.e. when we should not be scheduling any more txs for that slot.

there's two reasons:

- impl simplify and risk reduction (i.e. extracting `Bank::check_age()` is too dangerous without some rigourious code audit, which i want to avoid (esp considering nonces). Certainly, this is nice optimization opportunity. maybe future work).
- the recent blockhashes are exposed to runtime as a sysvar (`SysvarRecentB1ockHashes11111111111111111111`) in theory. _In theory_ here means this isn't the case due to the existence of `SysvarCache`, which is updated only at the start of slot. I don't want to rely on some peculiar behavior of the cache not to be fragiel. Again, this is fixable, but this will be of (distant) future work).

Alternatively, we can route the recentblockhash update via the scheduler like this:

```
enum SchedulerTask {
  ExecuteTransaction(txes...),
  UpdateRecentBlockhash(hash),
}
```

so that this particular blocking call can be consolidated into replaystage's `wait_for_completed_scheduler()`. but again, this is out-of-scope for now due to the need of auditing any potential affected access timings to the bank's recent blockhashes and its implications.